crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -85,6 +65,57 @@ pub struct CheckpointStore {
     watermarks: DBMap<CheckpointWatermark, (CheckpointSequenceNumber, CheckpointDigest)>,
 }
 
+pub trait PerEpochCheckpointStore {
+    fn get_epoch(&self) -> EpochId;
+
+    fn get_pending_checkpoints(
+        &self,
+    ) -> Vec<(CheckpointCommitHeight, (Vec<TransactionDigest>, bool))>;
+
+    fn get_pending_checkpoint(
+        &self,
+        index: &CheckpointCommitHeight,
+    ) -> Result<Option<(Vec<TransactionDigest>, bool)>, TypedStoreError>;
+
+    fn insert_pending_checkpoint(
+        &self,
+        index: &CheckpointCommitHeight,
+        transactions: &(Vec<TransactionDigest>, bool),
+    ) -> Result<(), TypedStoreError>;
+
+    fn delete_pending_checkpoints_return_batch(
+        &self,
+        commits: &[CheckpointCommitHeight],
+    ) -> Result<DBBatch, TypedStoreError>;
+
+    fn tx_checkpointed_in_current_epoch(
+        &self,
+        digest: &TransactionDigest,
+    ) -> Result<bool, TypedStoreError>;
+
+    fn record_checkpointed_tx_with_batch(
+        &self,
+        digest: &TransactionDigest,
+        checkpoint_seq: CheckpointSequenceNumber,
+        batch: DBBatch,
+    ) -> Result<DBBatch, TypedStoreError>;
```

Comment:
Introducing a trait where the Error type is fixed or internal implementation details are leaked (like the DBBatch type) tend to not be very flexible. Whats the purpose of this trait given there is only a single implementer.

Reply:
Good point

Reply:
Removed trait